### PR TITLE
settings: expose primitive callback registration

### PR DESCRIPTION
**Pull request by dt**
callers might want to just read an atomic int or something that the settings mechanism updates for them.
eventually we might want to polish use cases like that with dedicated wrappers, but we can start with
something simple like this and see what patterns end up proving useful.

### CODE REVIEW

**Comment by tamird** [comment]
> settings: simply settings test setup

simplify?

> it gets its own test binary, so this defer song-and-dance isn’t needed.

s/it/This package/

Also not sure why this is better.

LGTM modulo polluting the global namespace.

---

Reviewed 1 of 1 files at r1, 2 of 2 files at r2.
Review status: all files reviewed at latest revision, 1 unresolved discussion, some commit checks pending.

---

*[pkg/settings/cache.go, line 154 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/14866#-KhZ34rnccuSYz2YQMsE:-KhZ34roSfQrf6rrduup:bhqih9i) ([raw file](https://github.com/cockroachdb/cockroach/blob/8a1794a522207e9a23f97914eb3922d57dc081ea/pkg/settings/cache.go#L154)):*
> ```Go
> }
> 
> // RegisterCallback registers a func() f that will be called immediately and
> ```

"a func()" repeats the type. how about "registers f to be called"

inconsistent use of "f" here, it's spelled \`f\` below.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/14866#-:-KhZ2nG_Kn74q8Wa76kh:b-eqk6h9)*

**Comment by dt** [comment]
Reworded. This is better because then the vars are there at `init`, the way they would be if they were normal (e.g. just in the registry). It shouldn't pollute anything outside the settings_test binary though, right?

---

Review status: 0 of 2 files reviewed at latest revision, 1 unresolved discussion.

---

*[pkg/settings/cache.go, line 154 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/14866#-KhZ34rnccuSYz2YQMsE:-KhZ9pVYv77FlRw5AbFJ:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/8a1794a522207e9a23f97914eb3922d57dc081ea/pkg/settings/cache.go#L154)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

"a func()" repeats the type. how about "registers f to be called"

inconsistent use of "f" here, it's spelled \`f\` below.
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/14866#-:-KhZ9puuhb0Jql1-C1OD:b-mc8jo6)*

File: pkg/settings/cache.go [hunk 1]
```diff
@@ -138,4 +138,32 @@ func (u Updater) Apply() {
...
+	}
+	afterApply.Unlock()
+}
+
+var afterApply struct {
+	syncutil.Mutex
+	hooks []func()
+}
+
+// RegisterCallback registers `f`` to be called immediately and then after new
```
Comment (tschottdorf) [inline pkg/settings/cache.go]: s/``/`/

**Comment by dt** [comment]
Review status: 0 of 2 files reviewed at latest revision, 2 unresolved discussions.

---

*[pkg/settings/cache.go, line 154 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/14866#-KhZ1EX5tfetmHzLdLsG-r4-154:-KhbOBi3svEfQDpzi3t6:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/609ed277ea61536a65e50b383cda38c4cf07380a/pkg/settings/cache.go#L154)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary><blockquote>

s/``/`/
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/14866)*

**Comment by tamird** [comment]
Yeah, that's right. I still don't love it, because it will be more headache than anything should you decide to add another test.

---

Reviewed 1 of 2 files at r4, 1 of 1 files at r5.
Review status: all files reviewed at latest revision, 1 unresolved discussion, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/14866#-:-Khb_vaVa1kBb-ctNr_X:bezmjkz)*

**Comment by dt** [comment]
I'm not sure I'm fully understanding the issue you're worried about.. do you have an example? more importantly, do you have another approach you'd prefer to see?

**Comment by tamird** [comment]
If you add another test to the same package, you'll need to go back to the test-local state. You're making work for future you, so I'm ok with it. You have already my sign-off.

**Comment by dt** [comment]
well, conversely, the next test added might avoid having to reimplement setup/teardown of testdata if this works for it too, so I'm fine with calling it shared testdata for all tests in the package (all one at the moment) and revisiting that ifit becomes a problem.

**Comment by tamird** [comment]
👌

**Pull request merged**